zephyr/alloc.c

```diff
@@ -0,0 +1,220 @@
+// SPDX-License-Identifier: BSD-3-Clause
+/*
+ * Copyright(c) 2020 Intel Corporation. All rights reserved.
+ *
+ * Author: Liam Girdwood <liam.r.girdwood@linux.intel.com>
+ */
+
+#include <sof/lib/alloc.h>
+
+/* Zephyr includes */
+#include <device.h>
+#include <init.h>
+
+#ifndef CONFIG_KERNEL_COHERENCE
+#include <arch/xtensa/cache.h>
+#endif
+
+/*
+ * Memory - Create Zephyr HEAP for SOF.
+ *
+ * Currently functional but some items still WIP.
+ */
+
+#ifndef HEAP_RUNTIME_SIZE
+#define HEAP_RUNTIME_SIZE	0
+#endif
+
+/* system size not declared on some platforms */
+#ifndef HEAP_SYSTEM_SIZE
+#define HEAP_SYSTEM_SIZE	0
+#endif
+
+/* The Zephyr heap */
+#ifdef CONFIG_IMX
+#define HEAPMEM_SIZE		(HEAP_SYSTEM_SIZE + HEAP_RUNTIME_SIZE + HEAP_BUFFER_SIZE)
+/*
+ * Include heapmem variable in .heap_mem section, otherwise the HEAPMEM_SIZE is
+ * duplicated in two sections and the sdram0 region overflows.
+ */
+__section(".heap_mem") static uint8_t __aligned(64) heapmem[HEAPMEM_SIZE];
+#else
+#define HEAPMEM_SIZE		HEAP_BUFFER_SIZE
+#define HEAPMEM_SHARED_SIZE	(HEAP_SYSTEM_SIZE + HEAP_RUNTIME_SIZE + \
+				HEAP_RUNTIME_SHARED_SIZE + HEAP_SYSTEM_SHARED_SIZE)
+
+static uint8_t __aligned(PLATFORM_DCACHE_ALIGN)heapmem[HEAPMEM_SIZE];
+static uint8_t __aligned(PLATFORM_DCACHE_ALIGN)heapmem_shared[HEAPMEM_SHARED_SIZE];
+#endif
+
```

Comment:
@lyakh I don't think we should split this into 2 files, for imx and cavs - there is a lot of duplicated code.

I was thinking of keeping the `#ifdef CONFIG_IMX` at line 34, since we, for imx, need the heapmem variable in a different section.

The next part, the `#else` (starting with line 41) I believe it should be split into SMP and non-SMP.
For imx we only need non-SMP, APL has also 1 core, so there is no need for `heapmem` and `heapmem_shared` or `sof_heap` and  `sof_heap_shared`-  no need to split heap into shared and unshared.
The shared variables should be declared only for SMP. Please, correct me if I'm wrong.

I've tested only with `sof_heap` on imx and also @marc-hb verified on APL - see comment [here](https://github.com/thesofproject/sof/pull/4477#issuecomment-878548337).

In the rest of the file, where we use `sof_heap_shared`, should be under  `#ifdef CONFIG_SMP` - in functions like `statics_init()`, `rmalloc()`, `rfree()`.